Luis omega/architecture arweave (#6)
mlabs-haskell/DeNS: doc/Architecture.md

## Architecture - Arweave (Technical) (WIP)

It seems that the maximum amount of data we can upload paying the minimum amount possible is:

```bash
https://arweave.net/price/256000
212017846

> ```

Reviewer: What are the units of this? Is this bytes?

Author: yes, the last number is in bytes

> This price is given in Wilson, where a Wilson is related to Ar (the arweave currency) as:

Reviewer: ```suggestion
This price is given in Wilson, where a Wilson is related to Ar (the Arweave currency) as:
```
Iirc proper nouns should be capitalized?

```python
Ar = 1e12 wilson
```

> At today the price of an Ar is around `9.5` usd, this means that the minimum cost of a transaction is : `0.002014169537` usd.

Reviewer: ```suggestion
Today, the price of an Ar is around `9.5` USD, this means that the minimum cost of a transaction is : `0.002014169537` USD.
```

Reviewer: Why is this the minimum cost of the transaction? Is this just given in the documentation? Or are we calculating this based off of the previous discussion?

Author: I tried with a variable range of bytes, and it always was 212017846 for transactions below 250KB . The main hypothesis for Arweave to calculate price of storage is that the cost of it be less in the future.

And the maximum size of a transaction with such a price is above 250 KB.

Additionally some gateways allows the upload of certain amount of bytes for free.

> Those are the reason we choose to store every set of records in a particular zone file as a single transaction.

Reviewer: ```suggestion
Those are the reasons we choose to store every set of records in a particular zone file as a single transaction.
```

I don't feel convinced that these are good reasons for doing this. Can we expand on this some more / write this more clearly? 

Really spell out the details or something. A structure like:
- Discussion on zone files, how big they are, what we're expecting to actually have to store
- Remind the reader that this probably seems infeasible to do initially only on Cardano
- Talk about Arweave / how much it costs
- Then conclude that with Arweave is a reasonable ish solution to the storage problem.

Then, afterwards the discussion on how we actually plan to use Arweave is great followup.

Author: This is outdated after the meeting, will update soon.

### Initial Upload

> Initially we have a setup like this:

Reviewer: ```suggestion
Initially, we will assume we have the following zone files.
```
Iirc, commas should follow the adverb at the beginning of a sentence?

Maybe it would be good to recall to the reader _what_ a zone file is?

```bash
ZoneCOM.txt
ZoneNET.txt
.
.
.
ZoneSome.txt
```

> Every zone file will be split on smaller chunks of size `ORIGINAL_CHUNK_SIZE` (To be defined).

Reviewer: ```suggestion
Every zone file will be split into smaller chunks of size `ORIGINAL_CHUNK_SIZE` (to be defined).
```

Then using the services of a bundler we will upload the chunks in bulk.

> Arweave has a limit of transactions per block but, they allow the use of

Reviewer: ```suggestion
Arweave has a limit of transactions per block, but they allow the use of
```
Iirc, commas go before conjunctions like "but".

`bundle` transactions that allows the inclusion of up to `2^256` transactions
in every bundle.
Although the cost of upload can increase (to be investigated), this

> allows to a query by only downloading a single chunk instead of waiting for the download of

Reviewer: ```suggestion
allows one to a query by only downloading a single chunk instead of waiting for the download of
```

an entire zone file. Additionally, we believe that the increase cost won't be more than

> $100 usd and we expect it to be much less (to be investigated).

Reviewer: ```suggestion
$100 USD and we expect it to be much less (to be investigated).
```

The process of bundling will assign a unique transaction Id to the uploaded chunk

> This mean that at this point the reference inside a `DeNSValue` is

Reviewer: ```suggestion
This means that at this point the reference inside a `DeNSValue` is
```

the assigned id of the chunk.

This means that when someone tries to solve the ip of a domain, they
need to find the right domain inside the original chunk.

### Update of a set of records and upload of new records

> After the initial upload, every owner of a domain is responsibly of the

Reviewer: ```suggestion
After the initial upload, every owner of a domain is responsible for the
```

> update of they own records. With the use of a bundler, we can recommend

Reviewer: ```suggestion
update of their own records. With the use of a bundler, we can recommend
```

them to upload a single set of records in a single transaction. But
they may choose to not to bundle and just upload a single transaction
mixing records for other domains that they own. In such cases
the local dns would only update the local records corresponding to

> the domain that is being updated in cardano. This shouldn't prevent

Reviewer: ```suggestion
the domain that is being updated in Cardano. This shouldn't prevent
```

the user to refer in further updates of other domains to use

> the same transaction id for multiple domains.

Reviewer: What are we trying to say here?

This model means that the initial actor/mantainer of the network
would be able to do updates in bulk with low cost and to
discharge the update responsibility on a new owner in the future
with ease.

> 

Reviewer: How about we add a sentence or two  here to talk about _what we actually have to build_ for this. I think I'm getting a more clear idea of what you were envisioning

- [Interacting with ogmios](https://ogmios.dev/) which provides a convenient websockets to interact with the blockchain node, and we could update our database based on the information obtained by the websockets.

```mermaid
flowchart
    Cardano[Cardano node]-- Listens for events ---CardanoIndexer[Chain indexer] ;
    CardanoIndexer---SQL[(SQL database)];
    CardanoIndexer-- Query UTxOs ---TXBuilder[Tx builder];

>     Cardano-- Submit Txs ---TXBuilder;
> ```
> 
> where the lines denote communication, and the square boxes denote programs.

Reviewer: ````suggestion
    Cardano-- Submit Txs ---TXBuilder;
```

````
Let's get rid of the words below which explained what the diagram meant (since it's no longer lines and squares now)
